bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -341,12 +346,43 @@ private int readFromLogChannel(long entryLogId, BufferedReadChannel channel, Byt
         }
     };
 
+    private final  CacheLoader<Long, FileChannel> loader = new CacheLoader<Long, FileChannel> () {
+        public FileChannel load(Long entryLogId) throws Exception {
+
+            File file = findFile(entryLogId);
+            // get channel is used to open an existing entry log file
+            // it would be better to open using read mode
+            return new RandomAccessFile(file, "r").getChannel();
+        }
+    };
+    // close the file channel, when it was removed from cache
+    private final RemovalListener<Long, FileChannel> removalListener = new RemovalListener<Long, FileChannel>() {
+        public void onRemoval(RemovalNotification<Long, FileChannel> removal) {
+            FileChannel fc = removal.getValue();
+            if (null != fc) {
+                try {
+                    fc.close();
+                } catch (IOException e) {
+                    LOG.warn("Exception while closing channel for log file: {}", removal.getKey());
+                } finally {
+                    IOUtils.close(LOG, fc);
+                }
+            }
+        }
+    };
+
+    private final ExecutorService removeExecutor = Executors.newSingleThreadExecutor();
+
     /**
      * Each thread local buffered read channel can share the same file handle because reads are not relative
-     * and don't cause a change in the channel's position. We use this map to store the file channels. Each
-     * file channel is mapped to a log id which represents an open log file.
+     * and don't cause a change in the channel's position. We use this Guava cache to store the file channels. Each
+     * file channel is mapped to a log id which represents an open log file. When the file channel is removed/idle,
+     * it will be closed. temporally use 1h to relax replace after reading.
      */
-    private final ConcurrentMap<Long, FileChannel> logid2FileChannel = new ConcurrentHashMap<Long, FileChannel>();
+    private LoadingCache<Long, FileChannel> logid2FileChannel =  CacheBuilder.newBuilder()
+                .expireAfterAccess(1, TimeUnit.HOURS)
+                .removalListener(RemovalListeners.asynchronous(removalListener, removeExecutor))
```

Comment:
I don't think you need to use asynchronous removal listener here. because this would somehow change the behavior of `close` and `removeFromChannelsAndClose`. I would suggest using removalListener here, so the behavior will be the same as what it was before when a filechannel is removed and closed.